Add a framework level component leveraging Actions

GiudGiud opened the pull request:
refs #24103 
A few things to figure out

Requirements
- [ ] what characteristics of THM should be kept?
- [ ] inter-component dependency resolution. Either use multiple tasks, or ?

Names & syntax
- [x] find a good name for the base class. ComponentAction? ActionComponent? MooseComponent?
- [x] find a good name for the syntax. Can't be `[Components]`!
- [x] Current syntax is good right? Components/name/type=component_type

Workflow items
- [x] register additional tasks when registering new components that have them (instead of registering everything in Components/). This is because we are using meta actions
- [x] proper input auto-complete works? It should since syntax builds fine now

Code organization
- [ ] how to do different types of components. THM did by dimension: 1D/2D/file meshes. This would be kind of limiting since a lot of the things I m trying to do are dimension-agnostic. Maybe the solution is to not use inheritance? Or inheritance but with a 1D/2D/3D component helper/interface rather than a "main parent" class
- [ ] make a list of all the components we could want, and write down an inheritance tree for that

Types of components
- [ ] physics-components (components that define their own physics for convenience, for example SAM & THM ones)
- [x] components that spawn their own mesh
- [x] components from whole meshes output of mesh-generators (using saved mesh, we would have to be careful for duplicates if we do that)
- [x] components on a block of a mesh (done in the TMAP8 components)

Junctions
- [ ] how to make them modify the mesh for physics that require contiguous treatment
- [ ] then add the various UOs for THM-like junction?
- [ ] or no mesh-stitiching and just add sidesets for things like heat radiation?

GiudGiud committed:
Add an action to create component-action
refs #24103

GiudGiud committed:
Add a base class for components created with actions
refs #24103

GiudGiud committed:
Add an example component

GiudGiud committed:
Add a simple test adding two components

GiudGiud committed:
Add ability to use a CombinerGenerator as a final generator

GiudGiud committed:
Combine all the component meshes as the final mesh

GiudGiud committed:
Add more working attributes to component

GiudGiud committed:
Add basic translation and rotation to components

GiudGiud committed:
Add a simple test for adding physics on components

GiudGiud committed:
Add physics from components

GiudGiud committed:
More careful treatment of block restriction in base diffusion physics

GiudGiud committed:
Add option to add blocks to physics by Ids

GiudGiud committed:
Add a component that lives on a saved mesh

GiudGiud committed:
Transition to a cmg for extra features in cylinder component

GiudGiud committed:
Add ComponentName for components

GiudGiud committed:
Add a setup_component action as a test for dependencies

GiudGiud committed:
Add a component with inter-component depenencies

GiudGiud committed:
Add some basic APIs for component geometry

GiudGiud committed:
Register more missing apps

GiudGiud committed:
Add basic geometric info on component base

GiudGiud committed:
Add verbosity to component base class

GiudGiud committed:
Add basic geometry routines to cylinder component

GiudGiud committed:
Add an init component physics once physics are created and inited
Register setupMeshAction to systemComponents so we get a valid mesh

GiudGiud committed:
Avoid early geometry RMs with components

GiudGiud committed:
Better logic when all mesh gens are crated by components

GiudGiud committed:
Add a helper for dealing with physics from components

GiudGiud committed:
Add a new routine for 2D vector overlap checks

GiudGiud committed:
Add default postprocessing to diffusion physics

GiudGiud committed:
Hold a list of blocks on components

GiudGiud committed:
Improve error message for wrong block restriction
Add utilities for checking component type
and getting a component to physics

GiudGiud committed:
Add base class for multiple species physics diffusion

GiudGiud committed:
Add CG discretization for multiple species physics

GiudGiud committed:
Rename to ActionComponent

GiudGiud committed:
Clean up ActionComponents:
- move physics parameters all to PhysicsComponentHelper
- remove extra logs
- rename cylinder to cylindercomponent

GiudGiud committed:
Add syntax page to actionComponents

GiudGiud committed:
Add documentation to action-components and actions for creating these new components

GiudGiud committed:
Rename CombineDisjointComponents and MeshGeneratorComponent
for more descriptive names

GiudGiud committed:
Add component test definitions
fixup previous tests inputs for new component syntax
add test for saved mesh component

GiudGiud committed:
Allow multiple MGs per component

GiudGiud committed:
Support creating simulations with both ActionComponents and a Mesh block
Fix combiner errors in parallel

GiudGiud committed:
Use mooseInfo for more of messages from verbose
Properly keep track of blocks in CylinderComponent

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2290977/) on 8932805 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27593/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27593/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ed7323a589147b5ff07334f28050081b3c797b0b`

GiudGiud commented:
imo as long as there are no component ordering issues in creation / setup, then this is enough.
If there are, then I had a commit (removed for now) to introduce tasks and dependencies on-the-fly during construction

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2296900/) on f8b08c4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27593/site)

This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/actioncomponents/SavedMeshComponent.md:
```diff
@@ -0,0 +1,26 @@
+!alert! note
+There are large constraints when using a `SavedMeshComponent`:
+
+- the final mesh generator should be specified in the `[Mesh] block (this is true for most components)
```
```suggestion
- the final mesh generator should be specified in the `[Mesh]` block (this is true for most components)
```

GiudGiud committed:
Work around forcing users to create a Mesh block
Having a Mesh block is not a bad idea for the parameters it provides
Forcing it is not great. But the Mesh block is used for a lot of initialization, which we cannot:
- auto-create because of the THM components which use a THMMesh instead

Amend test hardcoded for first syntax name

GiudGiud committed:
Address documentation failures
- bad links
- sort tests under actioncomponents

GiudGiud committed:
Add a new register operation for Components, that registers
components to a bogus task so we can find them later

GiudGiud committed:
Add special logic the output of components, equivalent to the special logic
for outputting object parameters created by mooseObjectAction

GiudGiud committed:
Fix combiner generator errors when a distributed mesh does not have boundaries on all domains
refs #27646

GiudGiud commented:
This should be ready for review now. If it's too big I can remove the new multi-species diffusion physics. It's for TMAP but it fits fine in the scalar transport module imo. In the process of making it I found a lot of improvements for working with the single-species one

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2335712/) on f69c095 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>182f6e</th>
      <th colspan=4>#27593 f69c09</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.06%</td>
      <td>-0.00%</td>
      <td>78.53%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104585</td>
      <td>104876</td>
      <td>+291</td>
      <td>289</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18360</td>
      <td>18416</td>
      <td>+56</td>
      <td>79</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27593/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27593/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>182f6e</th>
      <th colspan=4>#27593 f69c09</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.63%</td>
      <td>84.63%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>16288</td>
      <td>16288</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2959</td>
      <td>2959</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27593/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27593/coverage/navier_stokes)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>182f6e</th>
      <th colspan=4>#27593 f69c09</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.49%</td>
      <td>42.20%</td>
      <td>-43.29%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>165</td>
      <td>165</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>28</td>
      <td>226</td>
      <td>+198</td>
      <td>198</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27593/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27593/coverage/scalar_transport)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27593/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 78.53% is less than the suggested 90.0%
- `scalar_transport` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2335712/) on f69c095 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 42.20% is less than the required 85.0%

GiudGiud committed:
Add a test for multi-species diffusion physics

GiudGiud committed:
Fix vector inputs in multi-species diffusion physics
Add some parameter size checks. #24103

pbehne committed:
Address code review:
- Removed trailing whitespace.
- clang format
- rename task

Co-authored-by: Patrick Behne <31858053+pbehne@users.noreply.github.com>
Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2355014/) on bd8b69c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27593/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27593/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3b0c2f9c72ad75fab4735f342aacb667eb51759b`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2355107/) on 95505d7 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 81.31% is less than the required 85.0%

pbehne commented on framework/include/utils/InputParametersChecksUtils.h:
```diff
@@ -353,9 +358,43 @@ InputParametersChecksUtils<C>::checkVectorParamsNoOverlap(
     for (const auto & value : forwardGetParam<std::vector<T>>(param))
       if (!unique_params.insert(value).second)
+      {
+        auto copy_params = param_vec;
+        copy_params.erase(std::find(copy_params.begin(), copy_params.end(), param));
```
What is the purpose of these two lines? To me, it looks like a local copy is made, a value is erased, and then the copy is never used.

GiudGiud replied:
it's to avoid saying
```
"Item '" xyz "' specified in vector parameter '" + parameter1 +
                            "' is also present in one or more of the two-D vector parameters '" +
                            "parameter1 parameter2 parameter3" +
                            "' in the inner vector of the same index, which is not allowed.");
```
which isnt clear to the user since parameter1's name shows up twice

joshuahansel replied:
I think @pbehne is suggesting you should have `stringify(copy_params)` instead of `stringify(param_vec)`

GiudGiud replied:
ah yes that's what I intended. good catch

pbehne commented on framework/include/utils/InputParametersChecksUtils.h:
```diff
@@ -353,9 +358,43 @@ InputParametersChecksUtils<C>::checkVectorParamsNoOverlap(
+InputParametersChecksUtils<C>::checkTwoDVectorParamsNoRespectiveOverlap(
+    const std::vector<std::string> & param_vec) const
+{
+  // Outer loop, each parameter is a vector of vector
```
Wouldn't each entry in a range-based loop of a vector of string just be a string? What trickery is going on here?

pbehne reviewed (commented):
So far, I have reviewed md and h files down to ActionComponent.C. I'll be on PL next week and will pick up where I left off when I return.

joshuahansel commented:
I can't remember, but I thought we were going with "ComponentAction" rather than "ActionComponent". The former indicates it's a type of Action, where the latter suggests it's a type of Component.

GiudGiud commented:
We want those to serve as components. The Action part of the name is just to differentiate them for now. I used Component as the suffix as this is what I'd like to use as the suffix for the derived classes

joshuahansel commented:
Ok, and to recap, is the plan to eventually convert the existing Component system to this, and then drop the old Component system and rename "ActionComponent" to "Component"?

GiudGiud commented:
in the theoretical world where we have funding to this next year, yes

joshuahansel commented on framework/include/actioncomponents/ActionComponent.h:
```diff
@@ -0,0 +1,92 @@
+  const std::vector<MeshGeneratorName> & meshGeneratorNames() const { return _mg_names; }
+
+  /// Returns the subdomains for the component mesh, if any
+  const std::vector<SubdomainName> & blocks() const { return _blocks; }
```
What are your thoughts on having something like `MeshedActionComponent` so that there aren't these interfaces for non-meshed components?

GiudGiud replied:
we really need to decide on this inheritance tree :O

joshuahansel replied:
This is my only suggestion to make to the tree. It can live without the change, but it's just going to be a little janky being able to call methods like `volume()` and such for 0-D components. What are your thoughts on the tree?

GiudGiud replied:
I would see 0D components have a volume in the same way that 1D and 2D ones have one. They represent something 3D

Same for a "boundary", we could have a nodeset of the 1 node act as the boundary, with a "surfaceArea" for the equivalent 3D element

My thoughts is no tree until we get forced into it. Just work with helpers and interfaces, which can be virtually inherited if needed

joshuahansel replied:
There'll probably be some abstract components where there is not an associated volume or such. But I'm fine with waiting - it won't be too hard to change later.

joshuahansel commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+                                     "Dimension of the cylinder. 1 for an (axial) 1D line, 2 for a "
+                                     "2D-RZ cylinder, and 3 for a 3D cylinder");
+  params.addRequiredRangeCheckedParam<Real>("radius", "radius>0", "Radius of the cylinder");
+  params.addRequiredRangeCheckedParam<Real>("height", "height>0", "Height of the cylinder");
```
I prefer "length" because "height" sounds vertical always.

GiudGiud replied:
good point

GiudGiud replied:
done

joshuahansel commented on test/tests/actioncomponents/gold/saved_mesh_component_out.e:
```diff
@@ -0,0 +1 @@
+two_same_components_out.e
```
So this is some feature of exodus files - that you can just list the name of another file?

GiudGiud replied:
this is how git shows the symlinks

GiudGiud commented on framework/doc/content/source/actioncomponents/SavedMeshComponent.md:
```diff
@@ -0,0 +1,25 @@
+There are large constraints when using a `SavedMeshComponent`:
+
+- the final mesh generator should be specified in the `[Mesh]` block (this is true for most components)
+- the mesh used by a `SavedMeshComponent` should be a saved mesh, using the `save_in` parameter of mesh generators
```
you're right. never hardcode param names in the doco

joshuahansel commented on framework/doc/content/source/actions/CreateMeshSetupActionsForComponents.md:
```diff
@@ -0,0 +1,13 @@
+!syntax description /ActionComponents/CreateMeshSetupActionsForComponents
+
+This `Action` creates the [SetupMeshAction.md] and the [SetupMeshCompleteAction.md] when the
+`[ActionComponents]` syntax is used. These two actions are normally on created if the `[Mesh]`
```
```suggestion
`[ActionComponents]` syntax is used. These two actions are normally only created if the `[Mesh]`
```

joshuahansel commented on framework/src/actions/AddActionComponentAction.C:
```diff
@@ -0,0 +1,66 @@
+  params.addClassDescription("Action responsible for creating component actions. A component "
+                             "action is a component derived from an Action base class");
+  params.addRequiredParam<std::string>(
+      "type", "A string representing the Component that will be built by this Action");
```
```suggestion
      "type", "Type of the ComponentAction to create with this Action");
```

joshuahansel commented on framework/src/actions/CreateMeshSetupActionsForComponents.C:
```diff
@@ -0,0 +1,60 @@
+#include "ActionFactory.h"
+#include "MooseObjectAction.h"
+
+registerMooseAction("MooseApp", CreateMeshSetupActionsForComponents, "component_meta");
```
I prefer "meta_action_component"

joshuahansel commented on modules/scalar_transport/src/physics/MultiSpeciesDiffusionCG.C:
```diff
@@ -0,0 +1,258 @@
+{
+  InputParameters params = MultiSpeciesDiffusionPhysicsBase::validParams();
+  params.addClassDescription(
+      "Discretizes a diffusion equation with the continuous Galerkin finite element method");
```
multi-species

GiudGiud replied:
```suggestion
      "Discretizes diffusion equations for several species with the continuous Galerkin finite element method");
```

GiudGiud commented on test/tests/actioncomponents/component_with_physics.i:
```diff
@@ -0,0 +1,99 @@
+    height = 1
+    position = '2 0 0'
+    direction = '0 0 1'
+    physics = 'added_from_component'
```
currently it would add another block (the component's) to the block restriction of the Physics

GiudGiud commented on test/tests/actioncomponents/two_same_components.i:
```diff
@@ -0,0 +1,35 @@
+[ActionComponents]
```
intent is different. might be the same though

GiudGiud replied:
the `component_with_physics.i` is a bigger input that tests working with physics. this one only checks we can have two components

GiudGiud commented on framework/doc/content/source/actioncomponents/SavedMeshComponent.md:
```diff
@@ -0,0 +1,25 @@
+# SavedMeshComponent
+
+The `SavedMeshComponent` is a geometrical component which lives on the mesh output by a mesh generator.
+The component's block restriction is set to the blocks of the saved mesh, and the saved mesh
```
a saved_mesh is a special case of MG mesh output. it  saves the MG output on the side for later use, notably it does not have to be the final mesh

GiudGiud commented on framework/src/actioncomponents/SavedMeshComponent.C:
```diff
@@ -0,0 +1,56 @@
+void
+SavedMeshComponent::setupComponent()
+{
+  // TODO: consider supporting the final mesh generator
```
MeshGeneratorComponent could be nice, but the saved_mesh part is crucial for the implementation at this point

GiudGiud replied:
let me try

GiudGiud replied:
ok done

GiudGiud commented on framework/doc/content/source/actions/CombineComponentsMeshes.md:
```diff
@@ -0,0 +1,15 @@
+!alert warning
+The components' meshes are currently combined with a [CombinerGenerator.md]. This means the
+mesh is not stitched at the interfaces. The user must specify the connection between components
+using either [boundary conditions](syntax/BCs/index.md) or dedicated [UserObjects](UserObjects/index.md).
```
since the meshes are not stitiched, interface kernels wont work. it has to be BCs or mortar-like treatment

GiudGiud commented on framework/doc/content/source/actions/CombineComponentsMeshes.md:
```diff
@@ -0,0 +1,15 @@
+!alert warning
+The components' meshes are currently combined with a [CombinerGenerator.md]. This means the
+mesh is not stitched at the interfaces. The user must specify the connection between components
+using either [boundary conditions](syntax/BCs/index.md) or dedicated [UserObjects](UserObjects/index.md).
```
```suggestion
using interface conditions. These can for example be set with [boundary conditions](syntax/BCs/index.md) or dedicated [UserObjects](UserObjects/index.md).
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2403623/) on 6968ad0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27593/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27593/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5e7bbe6b1c7733a3cf388bc914aa261fbc7ac700`

GiudGiud committed:
Change saved mesh component to be able to uxe the final mesh

GiudGiud commented on framework/doc/content/source/actioncomponents/SavedMeshComponent.md:
```diff
@@ -0,0 +1,25 @@
+!alert! note
+There are large constraints when using a `SavedMeshComponent`:
+
+- the final mesh generator should be specified in the `[Mesh]` block (this is true for most components)
```
basically only this one so far

GiudGiud commented on framework/include/actioncomponents/ActionComponent.h:
```diff
@@ -0,0 +1,92 @@
+class FEProblem;
+
+#define registerActionComponent(app_name, component_name)                                          \
+  registerMooseAction(app_name, component_name, "list_component")
```
init_component_physics is only for classes that inherit from the PhysicsComponentHelper
same 'add_mesh_generator' might not be used if the ActionComponent does not generate a mesh.

let me add a check, same as for Physics

GiudGiud commented on framework/src/actioncomponents/ActionComponent.C:
```diff
@@ -0,0 +1,53 @@
+    addPhysics();
+  // These tasks are there to match what the current combined Physics + Component do
+  // These combined components will likely still exist in the future, when it makes more
+  // sense to include the physics than to split it off into its own block
```
Maybe we ll add a new interface that adds the whole set of tasks, so we dont need that code in ActionComponent.C

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -78,6 +84,16 @@ class PhysicsBase : public Action, public InputParametersChecksUtils<PhysicsBase
+  void checkComponentType(const ActionComponent & component) const;
+  /// Most basic way of adding a component: simply adding the blocks to the block
+  /// restriction of the Physics. More complex behavior should be implemented by overriding
+  virtual void addComponent(const ActionComponent & component);
```
not so far. But we havent really done anything complex

GiudGiud committed:
Address code review:
- significantly expand action component documentation
- make cylinder discretization required
- add checks on tasks required by components
- add more explanations to use saved_mesh / final_mg

GiudGiud commented on framework/include/actioncomponents/CylinderComponent.h:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Cylinder on which one can define a Physics. The mesh is automatically created
+ */
+class CylinderComponent : public virtual ActionComponent, public PhysicsComponentHelper
```
No purpose here. But when the next helper comes along that also derives from ActionComponent, and one wants to use two helper that's when the virtual inheritance comes in handy

GiudGiud replied:
well no purpose isnt exactly true. Imagine I split here between components that use a Physics (parameter) and components that dont (only geometry, or define their own kernels)

then we decide to split between 1D 2D and 3D components for base classes. Then we end up having to have 3 derived classes for PhyisicsComponent and the others

joshuahansel replied:
So what's the breakdown of the virtual inheritance vs. interface approach? I think if you do go the virtual inheritance route, I feel like the `Helper` in `PhysicsComponentHelper` doesn't belong, since children are Physics components, not just helpers.

GiudGiud replied:
I can rename to PhysicsComponentBase.
I m not sure which approach is superior at this point

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -258,3 +275,14 @@ PhysicsBase::getCoupledPhysics(const bool allow_fail) const
+void
+PhysicsBase::checkComponentType(const ActionComponent & component) const
+{
+  // TODO figure out why dynamic_cast is not enough
```
let's try removing it again

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+  params.addRequiredParam<Point>("position", "Positional offset of the cylinder");
+  params.addRequiredParam<Point>("direction", "Direction of the cylinder");
+
+  params.addParam<unsigned int>("n_axial", 1, "Axial mesh of the cylinder");
+  params.addParam<unsigned int>("n_radial", 1, "Radial mesh of the cylinder");
+  params.addParam<unsigned int>("n_azimuthal", 1, "Azimuthal mesh of the cylinder");
```
done

GiudGiud commented on framework/doc/content/source/actioncomponents/ActionComponent.md:
```diff
@@ -0,0 +1,23 @@
+# ActionComponent
+
+`ActionComponents` are derived from [Actions](actions/Action.md). They are meant to facilitate the setup of
+complex simulations by splitting the definition of each part of the spatial systems involved.
```
added a sentence

GiudGiud commented on framework/include/actioncomponents/PhysicsComponentHelper.h:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * Helper class to help creating an entire physics
+ * TODO: Should this be an Interface rather than a Helper?
```
see my discussion with Patrick. Trying to avoid branching into a tree of classes yet

GiudGiud commented on framework/doc/content/source/actioncomponents/ActionComponent.md:
```diff
@@ -0,0 +1,23 @@
+
+`ActionComponent` is a base class for all `ActionComponents`.
+
+It provides several public APIs to interact with other systems such as:
```
done

GiudGiud commented:
Reviews should be addressed @joshuahansel @pbehne 
Sorry I missed a couple comments on Patrick's first round but I should have addressed everything now

joshuahansel commented on framework/src/actioncomponents/ActionComponent.C:
```diff
@@ -0,0 +1,72 @@
+  // task they use, and the base class calls the appropriate virtual member functions
+  mooseDoOnce(checkRequiredTasks());
+
+  // These tasks are conceptually what we image a mostly-geometrical component should do
```
```suggestion
  // These tasks are conceptually what we imagine a mostly-geometrical component should do
```

joshuahansel commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,125 @@
+  MooseEnum dims("0 1 2 3");
+  params.addRequiredParam<MooseEnum>("dimension",
+                                     dims,
+                                     "Dimension of the cylinder. 1 for an (axial) 1D line, 2 for a "
```
0 for what?

GiudGiud replied:
0 to avoid 0 being 1 in the mooseEnum

GiudGiud replied:
or 1 being 0 :D

GiudGiud commented on framework/include/utils/InputParametersChecksUtils.h:
```diff
@@ -353,9 +358,43 @@ InputParametersChecksUtils<C>::checkVectorParamsNoOverlap(
         forwardMooseError("Item '" + value + "' specified in vector parameter '" + param +
                           "' is also present in one or more of the parameters '" +
-                          Moose::stringify(param_vec) + "'. This is disallowed.");
+                          Moose::stringify(param_vec) + "', which is not allowed.");
```
```suggestion
                          Moose::stringify(copy_params) + "', which is not allowed.");
```

GiudGiud committed:
Doco fixes and use correct param string for error
Better dimension checks in CylinderComponent

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud commented on framework/doc/content/source/actioncomponents/CylinderComponent.md:
```diff
@@ -0,0 +1,25 @@
+!syntax description /ActionComponents/CylinderComponent
+
+The `CylinderComponent` is a simple [ActionComponent.md] which adds a cylindrical component to
+the mesh. The user selects the dimensionality of the cylinder using the [!param](/ActionComponents/CylinderComponent/dimension) parameter. For example, the `Cylinder` is
```
added

GiudGiud committed:
Rename PhysicsComponentHelper to Base to reflect its status as an ActionComponent class

pbehne commented on framework/src/meshgenerators/CombinerGenerator.C:
```diff
@@ -246,8 +246,10 @@ CombinerGenerator::copyIntoMesh(UnstructuredMesh & destination, const Unstructur
     std::set<subdomain_id_type> dest_ids;
-    source.subdomain_ids(source_ids, false);
-    destination.subdomain_ids(dest_ids, false);
+    source.subdomain_ids(source_ids, true);
+    destination.subdomain_ids(dest_ids, true);
+    mooseAssert(source_ids.size(), "Should have a subdomain");
+    mooseAssert(dest_ids.size(), "Should have a subdomain");
```
Why was this change needed?

GiudGiud replied:
that's a fix for distributed meshes which sometimes has subdomains not defined on one of the procs

pbehne reviewed (changes requested):
Since I have been slow to review this, I am releasing my next round of comments. I've reviewed down to Builder.C

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+  params.addRequiredParam<MooseEnum>("dimension",
+                                     dims,
+                                     "Dimension of the cylinder. 1 for an (axial) 1D line, 2 for a "
+                                     "2D-RZ cylinder, and 3 for a 3D cylinder");
```
It's not, I added that since then in the docstring

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+  params.addRequiredParam<Point>("position", "Positional offset of the cylinder");
+  params.addRequiredParam<Point>("direction", "Direction of the cylinder");
+
+  params.addParam<unsigned int>("n_axial", 1, "Axial mesh of the cylinder");
```
changed to Josh's suggestion, roughly the same

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+  params.addParam<unsigned int>("n_radial", 1, "Radial mesh of the cylinder");
+  params.addParam<unsigned int>("n_azimuthal", 1, "Azimuthal mesh of the cylinder");
+
+  params.addParam<std::vector<SubdomainName>>("block", "Block for the cylinder");
```
no plans, changed

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+  }
+  else
+  {
+    mooseError("3D is currently unsupported");
```
added another error earlier

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  // Place it in position
+  if (isParamValid("position"))
```
removed

GiudGiud commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -0,0 +1,120 @@
+    _mg_names.push_back(name() + "_translated");
+  }
+  // Rotate it as desired
+  if (isParamValid("direction"))
```
removed

GiudGiud commented on framework/src/actioncomponents/PhysicsComponentHelper.C:
```diff
@@ -0,0 +1,41 @@
+#include "PhysicsComponentHelper.h"
+
+InputParameters
+PhysicsComponentHelper::validParams()
```
only for registered classes technically. it does not show up anywhere if not registered

GiudGiud commented on framework/src/actioncomponents/PhysicsComponentHelper.C:
```diff
@@ -0,0 +1,41 @@
+InputParameters
+PhysicsComponentHelper::validParams()
+{
+  auto params = emptyInputParameters();
```
no wont hurt

GiudGiud commented on framework/src/actioncomponents/PhysicsComponentHelper.C:
```diff
@@ -0,0 +1,41 @@
+}
+
+PhysicsComponentHelper::PhysicsComponentHelper(const InputParameters & params)
+  : ActionComponent(params), _physics_names(getParam<std::vector<PhysicsName>>("physics"))
```
getting saved by the virtual inheritance here. better to be safe though

GiudGiud commented on framework/src/actioncomponents/SavedMeshComponent.C:
```diff
@@ -0,0 +1,56 @@
+  std::vector<SubdomainName> blocks_vec;
+  for (const auto bid : blocks)
+    blocks_vec.push_back(saved_mesh->subdomain_name(bid));
+}
```
fixed

GiudGiud committed:
Address Patrick's review:
- remove unneeded param checks
- properly save blocks in MGComponent
- fix error message

joshuahansel commented on framework/doc/content/source/actioncomponents/PhysicsComponentBase.md:
```diff
@@ -0,0 +1,19 @@
+# PhysicsComponentBase
+
+The `PhysicsComponentBase` is a base class designed to facilitate the user of [Physics](Physics/index.md)
```
```suggestion
The `PhysicsComponentBase` is a base class designed to facilitate the use of [Physics](Physics/index.md)
```

joshuahansel commented on framework/doc/content/source/actioncomponents/PhysicsComponentBase.md:
```diff
@@ -0,0 +1,19 @@
+!listing framework/src/actioncomponents/CylinderComponent.C start=registerMooseAction end=InputParameters
+
+!alert note
+This helper leverages virtual inheritance, it inherits the [ActionComponent.md] class virtually.
```
```suggestion
This helper leverages virtual inheritance: it inherits the [ActionComponent.md] class virtually.
```

joshuahansel commented on framework/src/actioncomponents/PhysicsComponentBase.C:
```diff
@@ -0,0 +1,43 @@
+    if (_verbose)
+      mooseInfoRepeated("Adding Physics '" + physics->name() + "' on component '" + name() +
+                        "' on blocks '" + Moose::stringify(_blocks) + "'");
+    physics->addBlocks(_blocks);
+    // NOTE: we could use addComponent instead.
+    // Or keep addComponent for adding Components from Physics
```
I think `physics->addComponent(*this)` makes sense. Is there any difference yet?

GiudGiud replied:
not at the moment. I ll change it, seems safer

GiudGiud commented on framework/src/actioncomponents/PhysicsComponentBase.C:
```diff
@@ -0,0 +1,43 @@
+    if (_verbose)
+      mooseInfoRepeated("Adding Physics '" + physics->name() + "' on component '" + name() +
+                        "' on blocks '" + Moose::stringify(_blocks) + "'");
+    physics->addBlocks(_blocks);
+    // NOTE: we could use addComponent instead.
+    // Or keep addComponent for adding Components from Physics
```
```suggestion
    physics->addComponent(*this);
```

GiudGiud committed:
Address Josh comment

- doco typo
- use addComponent for adding physics from component

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud committed:
Address test failures:
- fix input name
- add missing source & syntax pages

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2428533/) on ebc8bbc wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 81.31% is less than the required 85.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2429003/) on a344475 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 81.31% is less than the required 85.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2431644/) on f7f78c9 wanted to post the following:


The following coverage requirement(s) failed:
- `scalar_transport` coverage rate 81.31% is less than the required 85.0%

GiudGiud commented:
thank you both for the reviews!
